Prepare for -Wunsafe-buffer-usage-in-libc-call

geoffreygaren opened the pull request:
#### 9b61e37d88b779af79004c38d19ce30b172eed44
<pre>
Prepare for -Wunsafe-buffer-usage-in-libc-call
<a href="https://bugs.webkit.org/show_bug.cgi?id=286175">https://bugs.webkit.org/show_bug.cgi?id=286175</a>
<a href="https://rdar.apple.com/143158556">rdar://143158556</a>

Reviewed by Chris Dumez.

I enabled the setting locally and bracketed the failures with skip macros.

I also fixed a couple hyper-trivial cases.

We now have an Apple-port specific skip, so Apple engineers searching for
skips needs two regexes:

        WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN(?! //.*port)

and

        WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN.*[^-]Apple

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
* Source/JavaScriptCore/API/tests/ExecutionTimeLimitTest.cpp:
* Source/JavaScriptCore/API/tests/FunctionOverridesTest.cpp:
(testFunctionOverrides):
* Source/JavaScriptCore/API/tests/FunctionToStringTests.cpp:
(testFunctionToString):
* Source/JavaScriptCore/API/tests/JSObjectGetProxyTargetTest.cpp:
(testJSObjectGetProxyTarget):
* Source/JavaScriptCore/API/tests/MultithreadedMultiVMExecutionTest.cpp:
(startMultithreadedMultiVMExecutionTest):
* Source/JavaScriptCore/API/tests/testapi.mm:
(blockSignatureContainsClass):
(cacheFileInDataVault):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_templates.py:
* Source/JavaScriptCore/assembler/AssemblerCommon.h:
* Source/JavaScriptCore/assembler/Printer.h:
* Source/JavaScriptCore/b3/B3ValueRep.h:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/bytecode/PropertyCondition.h:
(JSC::PropertyCondition::PropertyCondition):
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromString):
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::ensureCanInitializeWithZeros):
* Source/JavaScriptCore/dfg/DFGVariableEvent.cpp:
* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::dumpBits):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::runFixpointPhase):
* Source/JavaScriptCore/heap/PreciseAllocation.cpp:
(JSC::PreciseAllocation::tryReallocate):
* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
(Inspector::ContentSearchUtilities::escapeStringForRegularExpressionSource):
* Source/JavaScriptCore/jit/CallFrameShuffler.cpp:
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeSeparatedWXHeaps):
* Source/JavaScriptCore/jit/ICStats.cpp:
(JSC::ICEvent::operator&lt; const):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/ExceptionFuzz.cpp:
(JSC::doExceptionFuzzing):
* Source/JavaScriptCore/runtime/FuzzerPredictions.cpp:
(JSC::readFileIntoString):
* Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h:
(JSC::GenericTypedArrayView&lt;Adaptor&gt;::tryCreate):
* Source/JavaScriptCore/runtime/IndexingType.cpp:
(JSC::dumpIndexingType):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
* Source/JavaScriptCore/runtime/Options.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::enableAssembler):
* Source/JavaScriptCore/tools/CompilerTimingScope.cpp:
* Source/JavaScriptCore/tools/FunctionOverrides.cpp:
(JSC::parseClause):
(JSC::FunctionOverrides::parseOverridesInFile):
* Source/JavaScriptCore/tools/Integrity.cpp:
* Source/JavaScriptCore/wasm/WasmValueLocation.h:
* Source/WTF/wtf/DateMath.cpp:
(WTF::findMonth):
(WTF::parseInt):
(WTF::parseLong):
* Source/WTF/wtf/FastMalloc.cpp:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/FilePrintStream.cpp:
(WTF::FilePrintStream::vprintf):
* Source/WTF/wtf/Int128.cpp:
(WTF::printInternal):
* Source/WTF/wtf/LockedPrintStream.cpp:
(WTF::LockedPrintStream::vprintf):
* Source/WTF/wtf/NumberOfCores.cpp:
(WTF::numberOfProcessorCores):
* Source/WTF/wtf/PrintStream.cpp:
* Source/WTF/wtf/PrintStream.h:
(WTF::printInternal):
* Source/WTF/wtf/SafeStrerror.cpp:
(WTF::safeStrerror):
* Source/WTF/wtf/ScopedPrintStream.h:
* Source/WTF/wtf/StackTrace.cpp:
(WTF::StackTracePrinter::dump const):
* Source/WTF/wtf/StringPrintStream.cpp:
* Source/WTF/wtf/SystemMalloc.h:
(WTF::SystemMalloc::zeroedMalloc):
(WTF::SystemMalloc::tryZeroedMalloc):
* Source/WTF/wtf/URL.cpp:
(WTF::URL::print const):
* Source/WTF/wtf/cocoa/CrashReporter.cpp:
(WTF::setCrashLogMessage):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
(WTF::readIDNAllowedScriptListFile):
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostHandleIndirectLog):
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
(WTF::LibraryPathDiagnosticsLogger::logError):
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::StrLength):
(WTF::double_conversion::StringBuilder::AddSubstring):
(WTF::double_conversion::StringBuilder::Finalize):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::Thread::initializePlatformThreading):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::operator==):
* Source/WTF/wtf/text/CString.cpp:
(WTF::operator==):
(WTF::operator&lt;):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::span8):
(WTF::span8IncludingNullTerminator):
(WTF::span):
(WTF::spanIncludingNullTerminator):
* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator&lt;&lt;):
* Source/WTF/wtf/text/icu/UTextProvider.cpp:
(WTF::uTextCloneImpl):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Clone):
(WTF::openLatin1UTextProvider):
* Source/WTF/wtf/threads/Signals.cpp:
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::copyShortASCIIString):
(WTF::localesMatch):
(WTF::createIterator):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::isSeparator):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::createICUConverter const):
(PAL::ErrorCallbackSetter::~ErrorCallbackSetter):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::isUndesiredAlias):
(PAL::WTF_REQUIRES_LOCK):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(accessibilitySearchCriteriaForSearchPredicate):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(parameterizedAttributeValueForTesting):
(formatForDebugger):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::showStyle):
* Source/WebCore/dom/ElementData.cpp:
(WebCore::UniqueElementData::UniqueElementData):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::showSubTreeAcrossFrame):
* Source/WebCore/dom/Position.cpp:
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::create):
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::VisiblePosition::debugPosition const):
* Source/WebCore/editing/VisibleSelection.cpp:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::showTreeWithIndent const):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::compactStackTrace):
* Source/WebCore/html/HTMLElement.cpp:
(dumpInnerHTML):
* Source/WebCore/html/ImageDataArray.cpp:
(WebCore::fillTypedArray):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::printLayoutTreeForLiveDocuments):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::setAlwaysUseUTF8):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::dump):
* Source/WebCore/page/FrameTree.cpp:
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::pageProperty):
* Source/WebCore/platform/FileHandle.cpp:
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::isMemoryAttributionDisabled):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::SincResampler):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::reset):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::getBufferSubDataImpl):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::prepare):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):
(WebCore::GraphicsContextGLCocoa::platformInitializeContext):
* Source/WebCore/platform/graphics/cocoa/TransformationMatrixCocoa.cpp:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData):
* Source/WebCore/platform/mac/CursorMac.mm:
(WebCore::cursor):
* Source/WebCore/platform/network/FormDataBuilder.cpp:
(WebCore::FormDataBuilder::appendFormURLEncoded):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::setCanSuspendLockedFileAttribute):
* Source/WebCore/platform/text/SegmentedString.cpp:
(WebCore::SegmentedString::advancePastSlowCase):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::advancePast):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::showLayerTree):
* Source/WebCore/rendering/RenderQuote.cpp:
(WebCore::subtagCompare):
(WebCore::quotesForLanguage):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::RenderBundle):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::RenderBundleEncoder):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
* Source/WebKit/NetworkProcess/Authentication/cocoa/AuthenticationManagerCocoa.mm:
(WebKit::AuthenticationManager::initializeConnection):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::deleteOldVersions):
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(encodeInvocationArguments):
(crashWithClassName):
(decodeInvocationArguments):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
(-[_WKRemoteObjectRegistry _sendInvocation:interface:]):
(-[_WKRemoteObjectRegistry _invokeMethod:]):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::valueFromNSValue):
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::create):
(WebKit::SandboxExtensionImpl::~SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::getSerializedFormat):
(WebKit::SandboxExtension::createHandleForTemporaryFile):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/Daemon/PCMDaemonEntryPoint.mm:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm:
(WebKit::handleXPCEndpointMessage):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatatypeTraits.h:
(WebKit::_WKWebExtensionSQLiteDatatypeTraits&lt;NSNumber::bind):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::initializeTimers):
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::stopTimerFired):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::compileAndCacheSandboxProfile):
(WebKit::tryApplyCachedSandbox):
* Source/WebKit/Shared/mac/WebMemorySampler.mac.mm:
(WebKit::WebMemorySampler::sampleSystemMalloc const):
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
(WebKit::WebKitSwiftLibrary):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::pathToPDFOnDisk):
* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluation.mm:
(WebKit::PDFScriptEvaluation::pdfDocumentContainsPrintScript):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityAttributeValue:forParameter:]):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::connectToService):
(WebPushTool::Connection::sendPushMessage):
(WebPushTool::Connection::getPushPermissionState):
(WebPushTool::Connection::requestPushPermission):
(WebPushTool::Connection::sendAuditToken):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
* Source/bmalloc/bmalloc/CompactAllocationMode.h:

Canonical link: <a href="https://commits.webkit.org/289133@main">https://commits.webkit.org/289133@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb87bbea23513c8ec3a4c5362bbf04386c34f6c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85399 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5127 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39824 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90520 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36434 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5242 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66395 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24203 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46677 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31824 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35502 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78383 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92063 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84433 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9329 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75015 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12968 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73397 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74137 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18471 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16898 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4784 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13331 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12704 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18152 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106823 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12527 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25744 "Failed to compile JSC") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16005 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14286 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
```diff
@@ -25,6 +25,9 @@
 #include "PASReportCrashPrivate.h"
+#include <wtf/Compiler.h>
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
Why aren't you using our compiler-specific pragmas?

geoffreygaren replied:
Good point. It looks like this is the first Apple-port-specific annotation in the codebase. I'll use "Apple ports", and then we can grep for `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN.*[^-]Apple`. ("non-Apple" is in use.)

cdumez commented on Source/JavaScriptCore/API/tests/MultithreadedMultiVMExecutionTest.cpp:
```diff
@@ -51,13 +51,17 @@ void startMultithreadedMultiVMExecutionTest()
         } \
     } while (false)
 
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
I'm surprising doing this around a `#define` does anything. I'd expected we either need to do it at the call sites or *within* the `CHECK` macro.

geoffreygaren replied:
Hmmm... I'm surprised too. I'll move it inside the macro.

cdumez commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -49,13 +49,17 @@
 #include <wtf/Threading.h>
 #include <wtf/threads/Signals.h>
 
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
Should we use the header-specific pragma?

geoffreygaren replied:
Did you mean WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN? I think WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN is slightly better here because in theory we could fix it.

cdumez replied:
Yes, I meant `WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN`. I thought that's what we used around headers. Yes we can fix it, but not here. Someone would have to go fix those headers. No big deal either way. I just preferred if this didn't show in my usual searches for `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN`

geoffreygaren replied:
Make sense. I can do that in a follow up.

cdumez commented on Source/WTF/wtf/FastMalloc.h:
```diff
@@ -684,7 +684,7 @@ void operator delete(T* object, std::destroying_delete_t, size_t size) { \
     object->T::~T(); \
     if (UNLIKELY(object->checkedPtrCountWithoutThreadCheck())) { \
-        memset(static_cast<void*>(object), 0, size); \
+        zeroSpan(unsafeMakeSpan(reinterpret_cast<char*>(object), size)); \
```
We should be able to use `zeroBytes(object);` since `sizeof(object) == size` per the `static_assert` above.

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderBundle.mm:
```diff
@@ -54,7 +54,9 @@ - (instancetype)initWithUsage:(MTLResourceUsage)usage renderStages:(MTLRenderSta
     , m_renderBundlesResources(resources)
     , m_descriptor(descriptor)
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
     , m_descriptorColorFormats(descriptor.colorFormats ? Vector<WGPUTextureFormat>(std::span { descriptor.colorFormats, descriptor.colorFormatCount }) : Vector<WGPUTextureFormat>())
```
I think `std::span { descriptor.colorFormats, descriptor.colorFormatCount }` could be replaced with `unsafeMakeSpan(descriptor.colorFormats, descriptor.colorFormatCount)` but maybe that is not the point to this PR, it is fine to leave as-is

geoffreygaren replied:
Yeah, I agree. I almost did that, then decided to leave this behind as a reminder to and add a span() accessor to the descriptor type.

geoffreygaren committed:
Prepare for -Wunsafe-buffer-usage-in-libc-call
https://bugs.webkit.org/show_bug.cgi?id=286175
rdar://143158556

Reviewed by Chris Dumez.

I enabled the setting locally and bracketed the failures with skip macros.

I also fixed a couple hyper-trivial cases.

We now have an Apple-port specific skip, so Apple engineers searching for
skips needs two regexes:

        WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN(?! //.*port)

and

        WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN.*[^-]Apple

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
* Source/JavaScriptCore/API/tests/ExecutionTimeLimitTest.cpp:
* Source/JavaScriptCore/API/tests/FunctionOverridesTest.cpp:
(testFunctionOverrides):
* Source/JavaScriptCore/API/tests/FunctionToStringTests.cpp:
(testFunctionToString):
* Source/JavaScriptCore/API/tests/JSObjectGetProxyTargetTest.cpp:
(testJSObjectGetProxyTarget):
* Source/JavaScriptCore/API/tests/MultithreadedMultiVMExecutionTest.cpp:
(startMultithreadedMultiVMExecutionTest):
* Source/JavaScriptCore/API/tests/testapi.mm:
(blockSignatureContainsClass):
(cacheFileInDataVault):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_templates.py:
* Source/JavaScriptCore/assembler/AssemblerCommon.h:
* Source/JavaScriptCore/assembler/Printer.h:
* Source/JavaScriptCore/b3/B3ValueRep.h:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/bytecode/PropertyCondition.h:
(JSC::PropertyCondition::PropertyCondition):
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromString):
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::ensureCanInitializeWithZeros):
* Source/JavaScriptCore/dfg/DFGVariableEvent.cpp:
* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::dumpBits):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::runFixpointPhase):
* Source/JavaScriptCore/heap/PreciseAllocation.cpp:
(JSC::PreciseAllocation::tryReallocate):
* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
(Inspector::ContentSearchUtilities::escapeStringForRegularExpressionSource):
* Source/JavaScriptCore/jit/CallFrameShuffler.cpp:
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeSeparatedWXHeaps):
* Source/JavaScriptCore/jit/ICStats.cpp:
(JSC::ICEvent::operator< const):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/ExceptionFuzz.cpp:
(JSC::doExceptionFuzzing):
* Source/JavaScriptCore/runtime/FuzzerPredictions.cpp:
(JSC::readFileIntoString):
* Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h:
(JSC::GenericTypedArrayView<Adaptor>::tryCreate):
* Source/JavaScriptCore/runtime/IndexingType.cpp:
(JSC::dumpIndexingType):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
* Source/JavaScriptCore/runtime/Options.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::enableAssembler):
* Source/JavaScriptCore/tools/CompilerTimingScope.cpp:
* Source/JavaScriptCore/tools/FunctionOverrides.cpp:
(JSC::parseClause):
(JSC::FunctionOverrides::parseOverridesInFile):
* Source/JavaScriptCore/tools/Integrity.cpp:
* Source/JavaScriptCore/wasm/WasmValueLocation.h:
* Source/WTF/wtf/DateMath.cpp:
(WTF::findMonth):
(WTF::parseInt):
(WTF::parseLong):
* Source/WTF/wtf/FastMalloc.cpp:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/FilePrintStream.cpp:
(WTF::FilePrintStream::vprintf):
* Source/WTF/wtf/Int128.cpp:
(WTF::printInternal):
* Source/WTF/wtf/LockedPrintStream.cpp:
(WTF::LockedPrintStream::vprintf):
* Source/WTF/wtf/NumberOfCores.cpp:
(WTF::numberOfProcessorCores):
* Source/WTF/wtf/PrintStream.cpp:
* Source/WTF/wtf/PrintStream.h:
(WTF::printInternal):
* Source/WTF/wtf/SafeStrerror.cpp:
(WTF::safeStrerror):
* Source/WTF/wtf/ScopedPrintStream.h:
* Source/WTF/wtf/StackTrace.cpp:
(WTF::StackTracePrinter::dump const):
* Source/WTF/wtf/StringPrintStream.cpp:
* Source/WTF/wtf/SystemMalloc.h:
(WTF::SystemMalloc::zeroedMalloc):
(WTF::SystemMalloc::tryZeroedMalloc):
* Source/WTF/wtf/URL.cpp:
(WTF::URL::print const):
* Source/WTF/wtf/cocoa/CrashReporter.cpp:
(WTF::setCrashLogMessage):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
(WTF::readIDNAllowedScriptListFile):
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostHandleIndirectLog):
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
(WTF::LibraryPathDiagnosticsLogger::logError):
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::StrLength):
(WTF::double_conversion::StringBuilder::AddSubstring):
(WTF::double_conversion::StringBuilder::Finalize):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::Thread::initializePlatformThreading):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::operator==):
* Source/WTF/wtf/text/CString.cpp:
(WTF::operator==):
(WTF::operator<):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::span8):
(WTF::span8IncludingNullTerminator):
(WTF::span):
(WTF::spanIncludingNullTerminator):
* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator<<):
* Source/WTF/wtf/text/icu/UTextProvider.cpp:
(WTF::uTextCloneImpl):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Clone):
(WTF::openLatin1UTextProvider):
* Source/WTF/wtf/threads/Signals.cpp:
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::copyShortASCIIString):
(WTF::localesMatch):
(WTF::createIterator):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::isSeparator):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::createICUConverter const):
(PAL::ErrorCallbackSetter::~ErrorCallbackSetter):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::isUndesiredAlias):
(PAL::WTF_REQUIRES_LOCK):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(accessibilitySearchCriteriaForSearchPredicate):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(parameterizedAttributeValueForTesting):
(formatForDebugger):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::showStyle):
* Source/WebCore/dom/ElementData.cpp:
(WebCore::UniqueElementData::UniqueElementData):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::showSubTreeAcrossFrame):
* Source/WebCore/dom/Position.cpp:
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::create):
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::VisiblePosition::debugPosition const):
* Source/WebCore/editing/VisibleSelection.cpp:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::showTreeWithIndent const):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::compactStackTrace):
* Source/WebCore/html/HTMLElement.cpp:
(dumpInnerHTML):
* Source/WebCore/html/ImageDataArray.cpp:
(WebCore::fillTypedArray):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::printLayoutTreeForLiveDocuments):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::setAlwaysUseUTF8):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::dump):
* Source/WebCore/page/FrameTree.cpp:
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::pageProperty):
* Source/WebCore/platform/FileHandle.cpp:
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::isMemoryAttributionDisabled):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::SincResampler):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::reset):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::getBufferSubDataImpl):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::prepare):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):
(WebCore::GraphicsContextGLCocoa::platformInitializeContext):
* Source/WebCore/platform/graphics/cocoa/TransformationMatrixCocoa.cpp:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData):
* Source/WebCore/platform/mac/CursorMac.mm:
(WebCore::cursor):
* Source/WebCore/platform/network/FormDataBuilder.cpp:
(WebCore::FormDataBuilder::appendFormURLEncoded):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::setCanSuspendLockedFileAttribute):
* Source/WebCore/platform/text/SegmentedString.cpp:
(WebCore::SegmentedString::advancePastSlowCase):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::advancePast):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::showLayerTree):
* Source/WebCore/rendering/RenderQuote.cpp:
(WebCore::subtagCompare):
(WebCore::quotesForLanguage):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::RenderBundle):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::RenderBundleEncoder):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
* Source/WebKit/NetworkProcess/Authentication/cocoa/AuthenticationManagerCocoa.mm:
(WebKit::AuthenticationManager::initializeConnection):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::deleteOldVersions):
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(encodeInvocationArguments):
(crashWithClassName):
(decodeInvocationArguments):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
(-[_WKRemoteObjectRegistry _sendInvocation:interface:]):
(-[_WKRemoteObjectRegistry _invokeMethod:]):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::valueFromNSValue):
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::create):
(WebKit::SandboxExtensionImpl::~SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::getSerializedFormat):
(WebKit::SandboxExtension::createHandleForTemporaryFile):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/Daemon/PCMDaemonEntryPoint.mm:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm:
(WebKit::handleXPCEndpointMessage):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatatypeTraits.h:
(WebKit::_WKWebExtensionSQLiteDatatypeTraits<NSNumber::bind):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::initializeTimers):
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::stopTimerFired):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::compileAndCacheSandboxProfile):
(WebKit::tryApplyCachedSandbox):
* Source/WebKit/Shared/mac/WebMemorySampler.mac.mm:
(WebKit::WebMemorySampler::sampleSystemMalloc const):
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
(WebKit::WebKitSwiftLibrary):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::pathToPDFOnDisk):
* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluation.mm:
(WebKit::PDFScriptEvaluation::pdfDocumentContainsPrintScript):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityAttributeValue:forParameter:]):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::connectToService):
(WebPushTool::Connection::sendPushMessage):
(WebPushTool::Connection::getPushPermissionState):
(WebPushTool::Connection::requestPushPermission):
(WebPushTool::Connection::sendAuditToken):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
* Source/bmalloc/bmalloc/CompactAllocationMode.h:

Canonical link: https://commits.webkit.org/289133@main
